File: Tree/C++/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        int level = 0;
        queue<TreeNode*> q;
        q.push(root);
        if (root == NULL) {
            return res;
        }
        while (!q.empty()) {
            res.push_back(vector<int>());
            int n = q.size();
            for (int i = 0; i < n; i++) {
                TreeNode *node = q.front();
                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
                res[level].push_back(node->val);
                q.pop();
            }
            if (level % 2 == 1) {
                reverse(res[level].begin(), res[level].end());
            }
            ++level;
        }
        return res;
    }
};

File: Microsoft/Golang/binary-tree-zigzag-level-order-traversal.go

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
    res := make([][]int, 0)
    if root == nil {
        return res
    }
    queue := make([]*TreeNode, 0)
    queue = append(queue, root)
    depth := 0
    for len(queue) > 0 {
        n := len(queue)
        level := make([]int, 0)
        for i := 0; i < n; i++ {
            level = append(level, queue[i].Val)
            if queue[i].Left != nil {
                queue = append(queue, queue[i].Left)
            }
            if queue[i].Right != nil {
                queue = append(queue, queue[i].Right)
            }
        }
        if depth % 2 == 1 {
            for j := 0; j < len(level)/2; j++ {
                level[j], level[len(level)-j-1] = level[len(level)-j-1], level[j]
            }
        }
        res = append(res, level)
        queue = queue[n:]
        depth++
    }
    return res
}
